src/proto/rpc.rs

```diff
@@ -0,0 +1,473 @@
+            }
+            RpcState::CredsFlavor => {
+                pstate.creds_flavor =
+                    read_u32(pstate, *byte, pstate.creds_flavor, RpcState::CredsLen)
+            }
+            RpcState::CredsLen => {
+                pstate.data_len = read_u32(pstate, *byte, pstate.data_len, RpcState::Creds);
+                if matches!(pstate.state, RpcState::Creds) && pstate.data_len == 0 {
+                    pstate.state = RpcState::VerifFlavor
+                }
+            }
+            RpcState::Creds => {
+                pstate.creds_data.push(*byte);
+                read_string(pstate, RpcState::VerifFlavor)
+            }
+            RpcState::VerifFlavor => {
+                pstate.verif_flavor =
+                    read_u32(pstate, *byte, pstate.verif_flavor, RpcState::VerifLen)
+            }
+            RpcState::VerifLen => {
+                pstate.data_len = read_u32(pstate, *byte, pstate.data_len, RpcState::Verif);
+                if matches!(pstate.state, RpcState::Verif) && pstate.cur_len == 0 {
+                    pstate.state = RpcState::End
+                }
+            }
+            RpcState::Verif => {
+                pstate.verif_data.push(*byte);
+                read_string(pstate, RpcState::End)
+            }
+            RpcState::End => {
+                pstate.payload.push(*byte);
+            }
+        };
+    }
+}
+
+fn get_nth_byte(value: u32, nth: u8) -> u8 {
+    let shift = 8 * (3 - nth);
+    ((value & (0xff << shift)) >> shift).try_into().unwrap()
+}
+
+fn push_u32(buffer: &mut Vec<u8>, data: u32) {
+    for i in 0..4 {
+        buffer.push(get_nth_byte(data, i));
+    }
+}
+
+fn push_string_pad(buffer: &mut Vec<u8>, data: String) {
+    let len: u32 = data.len().try_into().unwrap();
+    push_u32(buffer, len);
+    buffer.append(&mut data.as_bytes().to_vec());
+    if len % 4 != 0 {
+        for _ in 0..(4 - (len % 4)) {
+            buffer.append(&mut b"\x00".to_vec());
+        }
+    }
+}
+
+fn build_repl_portmap(pstate: ProtocolState, client_info: &ClientInfo) -> Vec<u8> {
+    let mut resp = Vec::<u8>::new();
+    match pstate.procedure {
+        // 0 => {}
+        3 => {
+            // getaddr / getport
+            // accepted state: 0 (RPC executed successfully)
+            resp.extend([0, 0, 0, 0]);
+            let localport = client_info.port.dst.unwrap();
+            match pstate.prog_version {
+                2 => {
+                    push_u32(&mut resp, localport as u32);
+                }
+                3 | 4 => {
+                    let addr = format!(
+                        "{}.{}.{}",
+                        client_info.ip.dst.unwrap(),
+                        localport >> 8,
+                        localport % 256
+                    );
+                    push_string_pad(&mut resp, addr);
+                }
+                _ => panic!("Wrong RPC version"),
```

Comment:
`panic!` means the execution of **masscanned** stops immediatly. Is it intended? Could this function return an error that could be handled or ignored by the caller?